ostd/src/task/processor.rs

```diff
@@ -231,3 +222,181 @@ impl Drop for DisablePreemptGuard {
 pub fn disable_preempt() -> DisablePreemptGuard {
     DisablePreemptGuard::new()
 }
+
+/// A guard for non-block section.
+///
+/// There are some cases when we'd like to run preemptibly but without being blocked.
+/// The `non_block_*` semantics are introduced to help catch sleep/block operations
+/// in those cases that would otherwise escape the classic atomic-mode check.
+pub struct NonBlockSectionGuard {
+    _privare: (),
+}
+
+impl NonBlockSectionGuard {
+    fn new() -> Self {
+        let non_block_count_ptr = current_task().unwrap().non_block_count().get();
+        // SAFETY: Nobody except the caller task itself will set or get the counter sequentially.
+        unsafe {
+            *non_block_count_ptr += 1;
+        }
+
+        Self { _privare: () }
+    }
+}
+
+impl Drop for NonBlockSectionGuard {
+    fn drop(&mut self) {
+        let non_block_count_ptr = current_task().unwrap().non_block_count().get();
+        // SAFETY: Nobody except the caller task itself will set or get the counter sequentially.
+        unsafe {
+            *non_block_count_ptr -= 1;
+        }
+    }
+}
+
+/// Enters non-block section.
+#[allow(unused)]
+pub fn non_block_start() -> NonBlockSectionGuard {
```

Comment:
Well, `non_block_start` [has very few users](https://elixir.bootlin.com/linux/v6.10.2/A/ident/non_block_start) even in the Linux kernel. As for Asterinas, I think this feature will simply have no users for the foreseeable future. It is generally a bad idea to add features that have no users.

I wonder what's the initial motivation for you to implement this feature. I think you should either
 - justify why this feature is so important and you need to add it now (just saying "Michal Hocko once gave a specific use case..." might not be enough), or
 - remove the feature from this PR.

Reply:
I tried to make the detection method complete to cover every possible case so I implemented the `non_block_*` semantics in this PR. I did noticed there are no downstream users now in OSTD/Asterinas so I attributed the function with `#[allow(unused)]. The reason why I kept the design is basically the two below:
- I was not sure if there are indeed no use cases in current codebase since I'm an expert who knows about implementation of every subsystems. I thought there might be someone before this PR who fixed a similar problem by adopting coarse-grained lock or disable_preempt, which harms performance.
- I thought it could be a reminder for future developers that there is a light-weight semantics other than `disable_preempt` that may fit their use cases.

Now since you stated there does not and will not exist a possible use case for non-block-section in OSTD/Asterinas, I'll remove the related semantics in next version.